src/query/phrase_query/phrase_weight.rs

```diff
@@ -45,10 +46,66 @@ impl PhraseWeight {
         &self,
         reader: &SegmentReader,
         boost: Score,
-    ) -> crate::Result<Option<PhraseScorer<SegmentPostings>>> {
-        let similarity_weight = self.similarity_weight.boost_by(boost);
-        let fieldnorm_reader = self.fieldnorm_reader(reader)?;
+    ) -> crate::Result<Option<Box<dyn PhraseScorer>>> {
+        match self.get_scorer_params(reader, boost)? {
+            Some((term_postings_list, similarity_weight, fieldnorm_reader)) => {
+                if self.slop == 0 {
+                    Ok(Some(Box::new(ExactPhraseScorer::new(
+                        term_postings_list,
+                        similarity_weight,
+                        fieldnorm_reader,
+                        self.scoring_enabled,
+                    ))))
+                } else {
+                    Ok(Some(Box::new(SlopPhraseScorer::new(
+                        term_postings_list,
+                        similarity_weight,
+                        fieldnorm_reader,
+                        self.scoring_enabled,
+                        self.slop,
+                    ))))
+                }
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn phrase_scorer_as_scorer(
+        &self,
+        reader: &SegmentReader,
+        boost: Score,
+    ) -> crate::Result<Option<Box<dyn Scorer>>> {
+        match self.get_scorer_params(reader, boost)? {
+            Some((term_postings_list, similarity_weight, fieldnorm_reader)) => {
+                if self.slop == 0 {
+                    Ok(Some(Box::new(ExactPhraseScorer::new(
+                        term_postings_list,
+                        similarity_weight,
+                        fieldnorm_reader,
+                        self.scoring_enabled,
+                    ))))
+                } else {
+                    Ok(Some(Box::new(SlopPhraseScorer::new(
+                        term_postings_list,
+                        similarity_weight,
+                        fieldnorm_reader,
+                        self.scoring_enabled,
+                        self.slop,
+                    ))))
+                }
+            }
+            _ => Ok(None),
+        }
+    }
```

Comment:
I did not find an elegant way to factorize this code which only differs on the return type.